### DISCUSSION TITLE
access ${cargo_build_dir} for additional byproducts (cbindgen headers)

### DISCUSSION THREAD

**Discussion by diresi** [corrosion-rs/corrosion#526] [Q&A] [answered]
I'm using v0.5.0 and cmake 3.20.2 and I'm probably missing something.
I have cargo custom builds that generate static libs and generate cbindgen headers from build.rs. Those do end up in `${cargo_build_dir}`.
Static libs are copied to the target directory (something outside of `$CMAKE_BINARY_DIR/corrosion`) alright, but that affects only .a files. Headers next to libs are left out.
Since I consider those headers to be byproducts as well, I'm trying to either copy them as well or point include paths to `${cargo_build_dir}` directly, which is something I seem to struggle with as I fail to read that path in the first place.

I can use genex to read `LOCATION`, but in cmake 3.20 there are limited options to strip the filename from it.
```
target_include_directories (mylib PRIVATE $<TARGET_PROPERTY:rustlib,LOCATION>)
```
I had the idea to explicitly use `_add_cargo_build()`, which gives me access to `${cargo_build_dir}` but then would either do another `cargo build` if provided with MANIFEST_PATH or fail due to MANIFEST_PATH missing.

I'm not sure about the included `corrosion_experimental_cbindgen()`, is it a good idea to play with it? It didn't produce any headers in my first attempts, and tbh I'd rather use the custom build I already have as it works with cargo directly as well.

So, does that even make sense?
Am I doing it all wrong?
What's the best way to get additional byproducts from cargo domain into cmake?

**Comment by jschwe** [comment, marked as answer]
`corrosion_experimental_cbindgen()` is definitly usable and will probably be renamed to just `corrosion_cbindgen()` when i have some spare time.

>Static libs are copied to the target directory (something outside of $CMAKE_BINARY_DIR/corrosion) alright, but that affects only .a files. Headers next to libs are left out.

Corrosion can't know what a user defined build script will do. Corrosion does know what `corrosion_experimental_cbindgen()` does.

> I had the idea to explicitly use `_add_cargo_build()`

Don't rely on implementation details of corrosion please.

If you want to use build.rs (and thats a perfectly viable option), you can use `corrosion_set_env_vars()` to set an environment variable with the location you would like to copy the headers to (e.g. `CBINDGEN_HEADER_DIR=${CMAKE_BINARY_DIR}/my_headers`, read it in your build.rs and copy the headers to the correct place from the build.rs.
In your CMake code you then know wich location your headers will appear in, so you can add include directories or install commands as necessary. You might need to explicit list your header files in CMake, mark them as generated files, and add a dependency on the CMake target that builds your crate, to inform CMake of how the files get generated.

**Comment by diresi** [comment]
> corrosion_experimental_cbindgen() is definitly usable and will probably be renamed to just corrosion_cbindgen() when i have some spare time.

That's good to know! Let me try harder to get some headers out of it :)

> If you want to use build.rs (and thats a perfectly viable option), you can use corrosion_set_env_vars() to set an environment variable with the location you would like to copy the headers to (e.g. CBINDGEN_HEADER_DIR=${CMAKE_BINARY_DIR}/my_headers, read it in your build.rs and copy the headers to the correct place from the build.rs.

That's exactly what I'm doing right now. It feels a bit odd, so I was looking for more elegant ways.

Thank you!

**Reply by diresi** [reply]
So I thought I can give an update.
After failing to inject build.rs-generated headers properly into cmake interface dependencies (I'm not too good at this) I now think that `corrosion_experimental_cbindgen` is much better. For example, it works :)

Thanks once more!
